### DISCUSSION TITLE
Plans for OpenDDS 3 and OpenDDS 4

### DISCUSSION THREAD

**Discussion by jrw972** [OpenDDS/OpenDDS#4183] [Announcements]
A few slides of the recent [townhall](https://github.com/OpenDDS/OpenDDS/discussions/4172) were devoted to OpenDDS 3, 4, and beyond.  I wanted to post that information here to encourage discussions and questions.

# What is left in OpenDDS 3?

There are two or three features that we want to finish before starting work on OpenDDS 4.
1. @iguessthislldo is working on building OpenDDS with CMake.
For OpenDDS 4, we will switch from an MPC-based build to a CMake-based build.
MPC support will be retained so that applications that use MPC will still build against a CMake-built OpenDDS.
2. @jrw972 is working on improving the configuration of OpenDDS.
Specifically, we want to provide parity between command-line options, the configuration file, and what can be changed programmatically.
We also want to allow configuration via environment variables, allow multiple configuration files, and provide a cleaner interface for programmatic configuration for all language bindings.
3. The third feature is improvements to `opendds_idl`.
We would like to fold `idl2jni` into `opendds_idl`, enhance `opendds_idl` to the point that `tao_idl` is unnecessary, and implement support for the new C++ language mapping.
Along the way, we plan to build out a plugin system that will hopefully ease the burden when developing new language bindings.
Some or all of these changes may be deferred until OpenDDS 4.

# Leading up to OpenDDS 4

The theme of OpenDDS 4 is removing things to enable future development (see the next section).

Specifically, 
* Upgrade compiler baseline to C++11/14
* Upgrade dependencies (the chief one being ACE/TAO)
* Remove JMS
* Remove internal statistics collection
* Remove Excel plugin
* Remove monitor
* Remove udp and multicast transports

Most of the features that are being removed are not being maintained and any project that is using them should migrate to a viable replacement like using the `rtps_udp` transport instead of the `udp` and `multicast` transports.
Thus, we think that most applications that work with OpenDDS 3 will work with OpenDDS 4.

# OpenDDS 4 Development

Once OpenDDS 4 has been released, we want to pursue the following improvements:

## Refactor/Rewrite DCPS, Transport, and Discovery

OpenDDS has accumulated a fair bit of technical debt in its 16-year existence.
One of the goals for OpenDDS 4 is to pay off (some) of this technical debt by refactoring and rewriting different layers and modules.
Ideally, this would be done in a way that makes it feasible for someone to develop a new transport or discovery with 1) no or minimal changes to the existing code and 2) without extensive knowledge of the rest of the system.
Our focus when refactoring and rewriting will be modularity and unit testing.

## Make RTPS the default discovery and transport

The `rtps_udp` transport is interoperable and supports advanced features like XTypes and DDS Security.
RTPS as a transport and a discovery will be the default in OpenDDS 4.
At the same time, we will deprecate the InfoRepo and start planning for its eventual removal.

We imagine that many users of OpenDDS use the InfoRepo with the `tcp` transport.
(If we are wrong, please let us know.)
Consequently, we believe that it is important to enable a similar usage pattern, that is, a participant connects to a discovery service via TCP and then to other participants via TCP.
For the discovery service, we plan to improve the RtpsRelay so that it has parity with the InfoRepo.
For TCP, we plan to develop an `rtps_tcp` transport.
This will not be spec-compliant because the spec is not ratified, but it will provide the same features as the existing `tcp` transport using RTPS messages.

Mostly likely, the removal of the InfoRepo and TAO will signal the arrival of OpenDDS 5.

## Modularization of the repository

Another goal for OpenDDS 4 is to reduce the amount of code in the repository by moving it to other repositories so we can focus on features important to most users.
For example, most users are probably not using the support for FACE which involves code in `opendds_idl` and a supporting library.
Moving support for FACE into another repository would accomplish the following:
1. It would improve build times since the code for FACE would not be compiled.
2. It would provide feedback on the popularity of FACE since it would be tracked independently of the main repository.
3. If everyone stops using FACE, we are not left with dead code in the main repository.

# Priorities

Another slide in the townhall laid out the priorities of the OpenDDS Foundation:
1. Client work
2. Increase the number of OpenDDS users
3. Increase the number of OpenDDs developers
4. Maintainability

This announcement has mainly focused on maintainability and we believe that pursuing maintainability will have a positive impact on client work and increase the number of users and developers.
However, it doesn't address the non-technical question of what features, documentation, examples, etc. are needed to attract new users and developers and make them successful with OpenDDS.
So, an open question to the community is how do we (the community) increase the number of OpenDDS users and developers?

**Comment by jwillemsen** [comment]
Have you consired [RIDL](https://github.com/RemedyIT/ridl) as replacement for TAO_IDL? RIDL is used for TAOX11/AXCIOMA and with RIDL there is no need to cross compile.

**Reply by jrw972** [reply]
Excellent question.  First, some background for other readers.

IDL compilation takes place in three ways in OpenDDS.  First, when the core libraries are being compiled, `tao_idl` is involved in compiling the IDL that defines spec-defined interfaces like participants, publishers, subscribers, etc.  When cross-compiling, you need a host version of `tao_idl` to perform this step.  Second, and the step that most users are familiar with, is compiling the IDL for DDS topics.  This process uses `tao_idl` to produce the language mapping, `opendds_idl` to produce type support, and `tao_idl` again to produce typed DataWriters and DataReaders.  See https://opendds.readthedocs.io/en/latest-release/devguide/getting_started.html#processing-the-idl.  Third, for certain language mappings, including C++11 and FACE, `opendds_idl` is used to produce the language mapping and then `tao_idl` produces typed DataWriters and DataReaders.  See https://opendds.readthedocs.io/en/latest-release/devguide/opendds_idl.html#opendds-idl-using-the-idl-to-c-11-mapping.

A drop-in replacement for `tao_idl` would mean that we would not have to compile the backend for `tao_idl` but we would still need the frontend for `opendds_idl`.  Along the same lines, the cross-compilation issue would not be solved.

Ideally, we would have a single tool with the combined functionality of `opendds_idl` and `tao_idl` with pluggable backends.  Thus, I think the question is, "What about RIDL as a replacement for `opendds_idl` and `tao_idl`?"  I see two technical issues that need to be addressed with respect to RIDL and other candidates.  First, the front-end of `tao_idl` supports annotations which are necessary for XTypes.  RIDL is advertised as supporting IDL 3.5 which does not have annotations.  Second, there is shared code between `opendds_idl` and the core libraries.  Again, this is XTypes-related and concerns the generation and serialization of TypeObjects.  A change in architecture could resolve this dependency and writing a wrapper to expose the necessary functionality is possible.  However, for the time being, this code reuse is valuable.  Another issue is long-term maintainability and recruiting the next generation of OpenDDS developers.  Since most of OpenDDS is written in C++, the same pool of developers can work on both `opendds_idl` and the the core libraries.  If we choose to implement `opendds_idl` in another language, that may create friction.  On the other hand, if someone wants to contribute a rewrite of `opendds_idl`, then I think we would seriously have to consider the possibility.

**Reply by jwillemsen** [reply]
RIDL does support IDL4 annotations, that was lacking in the readme, just added that.
